heron/scheduler-core/src/java/com/twitter/heron/scheduler/UpdateTopologyManager.java

```diff
@@ -0,0 +1,162 @@
+      throws ExecutionException, InterruptedException {
+    String topologyName = Runtime.topologyName(runtime);
+    PackingPlan existingPackingPlan = deserializer.fromProto(existingProtoPackingPlan);
+    PackingPlan proposedPackingPlan = deserializer.fromProto(proposedProtoPackingPlan);
+
+    Integer proposedContainerCount = proposedPackingPlan.getContainers().size();
+    Integer existingContainerCount = existingPackingPlan.getContainers().size();
+    Integer containerDelta = proposedContainerCount - existingContainerCount;
+
+    assertTrue(proposedContainerCount > 0,
+        "proposed packing plan must have at least 1 container %s", proposedPackingPlan);
+    assertTrue(containerDelta == 0 || scalableScheduler.isPresent(),
+        "Topology change requires scaling containers by %s but scheduler does not support "
+        + "scaling, aborting. Existing packing plan: %s, proposed packing plan: %s",
+        containerDelta, existingPackingPlan, proposedPackingPlan);
+
+    SchedulerStateManagerAdaptor stateManager = Runtime.schedulerStateManagerAdaptor(runtime);
+
+    // assert found PackingPlan is same as existing PackingPlan.
+    PackingPlans.PackingPlan foundPackingPlan = stateManager.getPackingPlan(topologyName);
+    assertTrue(foundPackingPlan.equals(existingProtoPackingPlan),
+        "Existing packing plan received does not equal the packing plan found in the state "
+        + "manager. Not updating updatedTopology. Received: %s, Found: %s",
+        existingProtoPackingPlan, foundPackingPlan);
+
+    // fetch the topology, which will need to be updated
+    TopologyAPI.Topology updatedTopology = stateManager.getTopology(topologyName);
+    Map<String, Integer> proposedComponentCounts = proposedPackingPlan.getComponentCounts();
+    for (String componentName : proposedComponentCounts.keySet()) {
+      Integer parallelism = proposedComponentCounts.get(componentName);
+      updatedTopology = mergeTopology(updatedTopology, componentName, parallelism);
+    }
+
+    // request new resources if necessary. Once containers are allocated we should make the changes
+    // to state manager quickly, otherwise the scheduler might penalize for thrashing on start-up
+    if (containerDelta > 0 && scalableScheduler.isPresent()) {
+      scalableScheduler.get().addContainers(containerDelta);
+    }
+
+    // update parallelism in updatedTopology since TMaster checks that
+    // Sum(parallelism) == Sum(instances)
+    logFine("Deleted existing Topology: %s", stateManager.deleteTopology(topologyName));
+    logFine("Set new Topology: %s", stateManager.setTopology(updatedTopology, topologyName));
+
+    // update packing plan to trigger the scaling event
+    logFine("Deleted existing packing plan: %s", stateManager.deletePackingPlan(topologyName));
+    logFine("Set new PackingPlan: %s",
+        stateManager.setPackingPlan(proposedProtoPackingPlan, topologyName));
+
```

Comment:
There are 4 state manager interactions above. Failure in any one can result in state corruption and existing topology fail. In my opinion update operation must not negatively impact existing deployment. So I think state update (both packingPlan and topology) should be a transaction. Although curator (ZK) does support atomic batch operations, other state managers might not support it. I am thinking if we can implement a "current/active version marker" pattern. An update operation will look like this

1. Add new topology with a version number, say topology-v2
2. Add new packing plan, packing-v2
3. Atomically update topology's current/active version marker to v2
4. Asynchronously delete topology-v1 and packing-v1
5. consumers detect version marker change and read new packing and topology

What do you think?